### PR TITLE
Add support for 4-servo CCPM helicopters

### PR DESCRIPTION
**Pull request by fhedberg**
Adds support for 4-servo CCPM helicopters.

@R-Lefebvre Not sure if this is the best way to handle both 3 and 4 servo swashes conditionally, in order not to take up unnecessary channels. Thoughts on this?

### CODE REVIEW

File: ArduCopter/Parameters.h [hunk 1]
```diff
@@ -162,10 +162,11 @@ class Parameters {
         //
         // 80: Heli
         //
+        k_param_heli_servo_4 = 79,
         k_param_heli_servo_1 = 80,
         k_param_heli_servo_2,
         k_param_heli_servo_3,
-        k_param_heli_servo_4,
```
Comment (magicrub) [inline ArduCopter/Parameters.h]: I'm not sure this is going to work

Comment (fhedberg) [inline ArduCopter/Parameters.h]: How so?

Comment (magicrub) [inline ArduCopter/Parameters.h]: data in k_param_heli_servo_4 will be lost/corrupted, and posisble auto-populated into servo_yaw thus never allowing the defaults to be loaded. Better to create the _servo_yaw at 79. Also, please either change or remove the // 80: Heli label.

File: libraries/AP_Motors/AP_MotorsHeli.h [hunk 2]
```diff
@@ -73,7 +73,7 @@ class AP_MotorsHeli : public AP_Motors {
     };
 
     // init
-    void Init();
+    virtual void Init();
```
Comment (magicrub) [inline libraries/AP_Motors/AP_MotorsHeli.h]: make pure?

File: ArduCopter/Parameters.h (hunk 1 above)
Comment (fhedberg) [inline ArduCopter/Parameters.h]: The purpose is actually just that - in essence renaming the forth servo to servo_yaw and inserting a new servo 4. That way we won't loose any params when upgrading from older versions.

File: libraries/AP_Motors/AP_MotorsHeli.h (hunk 2 above)
Comment (fhedberg) [inline libraries/AP_Motors/AP_MotorsHeli.h]: Not sure I understand why?

**Comment by R-Lefebvre** [comment]
Do you have an imminent need for this?  I'm just trying to avoid adding more and more complexity unless there's a reason for it.  Ideally, I'd prefer if 4-servo swash was a compile-time option.  Although I know that other people want to get away from compile-time choices.  I don't like driving complexity onto the standard users.

We've been adding a lot of complexity lately for very rare use-cases (ie: groups of 1).

**Comment by fhedberg** [comment]
I do, and figured there was others that would find it useful, hence the PR. But am fine with maintaining it out of master.

**Comment by 0n3-70uch** [comment]
Hello,

I used your Code changes and implemented these changes into the new ArduCopter 3.4.3 code. The four servo swash Code works fine... Thank you.

Here is a short Video:
https://youtu.be/SySpWeiPe2c

Greetings Fabi

**Comment by fhedberg** [comment]
@0n3-70uch Awesome! Great to hear.

**Comment by 0n3-70uch** [comment]
Yeah but there is another big problem. Read here:
https://github.com/ArduPilot/ardupilot/issues/5396

**Comment by maxdrougge** [comment]
Hello. I'm quite (very) new to this forum and using pixhawk/ardupilot. I'm wondering if version 3.4.3 which supports 4 servo swash plate is stable and "safe" to fly with? If not, is there another improved version?

**Comment by 0n3-70uch** [comment]
Hello for me the 4 servo swash code is working great. But its not official at the moment...

**Comment by maxdrougge** [comment]
Alright. If I may ask, which is the latest software version which supports 4 servo swash plate?

**Comment by 0n3-70uch** [comment]
There is no official arducopter software with the support for 4 swash servos. Only some private builds based on fhedberg code.

**Comment by maxdrougge** [comment]
I understand. Thank you for your swift response. What hardware is most suitable to run this software (which platform are you using)?

**Comment by 0n3-70uch** [comment]
Your are welcome :-). I'm using the normal Pixhawk aka px4v2. But I think the code is also runnung on other hardware systems.

**Comment by maxdrougge** [comment]
I live in Sweden and found this one: http://www.rctech.se/index.php?route=product/product&path=108&product_id=453
Is this a suitable choice?

**Comment by 0n3-70uch** [comment]
I never tested the code on this one but i thik its a good platform. It's a bit newer and more powerful than the old Pixhawk.

**Comment by maxdrougge** [comment]
Ok. Last question and maybe a stupid question. Where do I download the software file(s) that I should upload to the Pixhawk? I can't seem to find it.

**Comment by 0n3-70uch** [comment]
Ok thats not so easy. You need to compile your own code. Here is a tutorial: http://ardupilot.org/dev/docs/building-px4-with-make.html

**Comment by maxdrougge** [comment]
Thanks I will try that. However, which version should I make if I want the 4 servo swash plate support?

**Comment by 0n3-70uch** [comment]
You can implement the function into every version. fhedberg did that for version 3.3 but you can also implement the 4 servo swash code into version 3.5.3.

**Comment by maxdrougge** [comment]
Ok. So what I need to do is pull version 3.5.3 and do the changes in the code according to the "Files changed" section in this post?

**Comment by 0n3-70uch** [comment]
Yeah right. But better you use version 3.3.3 because there is a problem with the attitude controller since version 3.4.

**Comment by maxdrougge** [comment]
Alright. Thanks for the support. Is there any way you would let me contact you via direct message or something if I run into problems or have a few questions once I try to get this going?

**Comment by 0n3-70uch** [comment]
Haha yeah you can contact me here: https://www.coptertec.de/kontakt/

**Comment by maxdrougge** [comment]
Does this support PixHawk v2.1? If not, is there a version with 4 swash-plate support that does?

**Pull request closed without merging**